tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,384 @@
+SeekEstablishesInv2 == [][\A offset \in Offset: Seek(offset) => Inv2']_vars
+Write1Correct == [][\A symbol \in Symbols: Write1(symbol) => RAF!Write(SeqToArray(<<symbol>>))]_vars
+Read1Correct == [][\A symbol \in Symbols: Read1(symbol) => RAF!Read(SeqToArray(<<symbol>>))]_vars
+WriteAtMostCorrect == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): WriteAtMost(data) => \E written \in 1..len: RAF!Write(ArraySlice(data, 0, written))]_vars
+ReadCorrect == [][\A len \in 1..MaxOffset: \A data \in Array(Symbols, len): Read(data) => RAF!Read(data)]_vars
+
+-------------------------------------------------------------------------------
+\* Model checking helper definitions
+
+Symmetry == Permutations(Symbols)
+
+Alias == [
+    \* constants
+    BuffSz            |-> BuffSz,
+    MaxOffset         |-> MaxOffset,
+
+    \* regular vars
+    dirty             |-> dirty,
+    length            |-> length,
+    curr              |-> curr,
+    lo                |-> lo,
+    buff              |-> buff,
+    diskPos           |-> diskPos,
+    file_content      |-> file_content,
+    file_pointer      |-> file_pointer,
+
+    \* abstract vars
+    abstract_contents |-> LogicalFileContent]
+
+===============================================================================
+
+--------------------------- MODULE RandomAccessFile ---------------------------
+
+EXTENDS Naturals, Sequences, Common
+
+VARIABLES
+    file_content,
+    file_pointer
+
+vars == <<file_content, file_pointer>>
+
+TypeOK ==
+    /\ file_content \in ArrayOfAnyLength(Symbols \union {ArbitrarySymbol})
+    /\ ArrayLen(file_content) <= MaxOffset
+    /\ file_pointer \in Offset
+
+Init ==
+    /\ file_content = EmptyArray
+    /\ file_pointer = 0
+
+Seek(new_offset) ==
+    /\ new_offset \in Offset
+    /\ file_pointer' = new_offset
+    /\ UNCHANGED <<file_content>>
+
+SetLength(new_length) ==
+    /\ file_content' = TruncateOrExtendFile(file_content, new_length)
+    /\ file_pointer' = Min(file_pointer, new_length)
+
+Read(output) ==
+    /\ output = ArraySlice(file_content, file_pointer, Min(file_pointer + ArrayLen(output), ArrayLen(file_content)))
+    /\ file_pointer' = file_pointer + ArrayLen(output)
+    /\ UNCHANGED <<file_content>>
+
+Write(data) ==
+    /\ file_pointer + ArrayLen(data) <= MaxOffset
+    /\ file_content' = WriteToFile(file_content, file_pointer, data)
+    /\ file_pointer' = file_pointer + ArrayLen(data)
+
+Next ==
+    \/ \E offset \in Offset:
+        \/ Seek(offset)
+        \/ SetLength(offset)
+    \/ \E len \in 1..MaxOffset: \E data \in Array(Symbols, len):
+        \/ Write(data)
+        \/ Read(data)
+
+Spec == Init /\ [][Next]_vars
+
+===============================================================================
+
+-------------------------------- MODULE Common --------------------------------
+
+EXTENDS Naturals, Sequences
+
+CONSTANTS
+    Symbols,
+    ArbitrarySymbol,
```

Comment:
Seems like the purpose and value of `ArbitrarySymbol` is worth calling out (maybe here or in `TruncateOrExtendFile` or there might be some better place).  If I understand it correctly, it seems that its to reign in state space explosion by mixing in a bit of symbolic checking?  It appears as though on could consider all possible extensions of a file using `Symbols`, but the downside of this is that we're going to potentially explore many states.  Whether I understood correctly or not, it's probably worth explaining.